skbio/io/_registry.py

```diff
@@ -90,28 +89,13 @@ def register_sniffer(format):
     """
     def decorator(sniffer):
         if format in _sniffers:
-            raise DuplicateRegistrationError(msg="'%s' already has a sniffer."
-                                             % format)
-
-        def wrapped_sniffer(fp, mode='U', **kwargs):
-            with open_file(fp, mode) as fh:
-                # The reason we do a copy is because we need the sniffer to not
-                # mutate the orginal file while guessing the format. The
-                # naive solution would be to seek to 0 at the end, but that
-                # would break an explicit offset provided by the user. Instead
-                # we create a shallow copy which works out of the box for
-                # file-like object, but does not work for real files. Instead
-                # the name attribute is reused in open for a new filehandle.
-                # Using seek and tell is not viable because in real files tell
-                # reflects the position of the read-ahead buffer and not the
-                # true offset of the iterator.
-                if hasattr(fh, 'name'):
-                    cfh = open(fh.name, fh.mode)
-                else:
-                    cfh = copy.copy(fh)
-                    cfh.seek(0)
+            raise DuplicateRegistrationError(
+                msg="'%s' already has a sniffer." % format)
+
+        def wrapped_sniffer(fp, mode='r', binary=binary, **kwargs):
```

Comment:
We should be able to drop `mode` as a registry-facing kwarg, and I don't think we actually want users to be able to override `binary` either, so you should be able to just leave that information in the closure.